src/codegen/compile.ml

```diff
@@ -1529,15 +1555,31 @@ module Tagged = struct
     | StableSeen -> 0xffffffffl
 
   (* The tag *)
-  let header_size = 1l
+  let header_size = 2l
   let tag_field = 0l
+  let forwarding_pointer_field = 1l
 
-  (* Assumes a pointer to the object on the stack *)
-  let store tag =
+  let alloc env size tag =
+    let (set_object, get_object) = new_local env "new_object" in
+    Heap.alloc env size ^^
+    set_object ^^ get_object ^^
+    compile_unboxed_const (int_of_tag tag) ^^
+    Heap.store_field tag_field ^^
+    get_object ^^ (* object pointer *)
+    get_object ^^ (* forwarding pointer *)
+    Heap.store_field forwarding_pointer_field ^^
+    get_object
+    
+  let load_forwarding_pointer env =
+    Heap.load_field forwarding_pointer_field
+
+  let store_tag env tag =
+    load_forwarding_pointer env ^^
     compile_unboxed_const (int_of_tag tag) ^^
     Heap.store_field tag_field
-
-  let load =
+    
+  let load_tag env =
+    load_forwarding_pointer env ^^
```

Comment:
This is an interesting question. I was also first thinking of only resolving it in sanity check mode. However, I believe it could be relevant during stabilization where the tags are changed to store the serialization sharing status (`StableSeen`, serialization offset). As serialization may start on non-completed evacuation and see old references, it would be important to also forward the tag accesses (load and store). Maybe, this could be simplified (or only done during serialization)?

Reply:
Oh, I forgot about that hack during stabilization, well observed!